Elc vs MMM2D fixes, tests and docu

KonradBreitsprecher opened the pull request:
Part of #1559 

 - Testcases that compare elc vs mmm2d for inert, dielectric and constant potential BCs for neutral and noneutral systems. The tests pulls a test charge through a fixed configuration of charges and compares forces and energies between mmm2d and elc. The testcase is pretty slow with -np 4 (~2 x 23s)
- Almost compliant interface for setting dielectric contrast and constant potential for the two.
- Sanity check for elc: nonneutral + nonmetallic BCs not allowed
- Sanity check for mmm2d: verlet lists not allowed
- Extended docu covering nonneutral systems and dielectrics for elc.
- Fixed constant potental energy sign issue in elc similar to #1568. Covered by the testcase.
- Fixed MMM2d const pot MPI bug.

RudolfWeeber committed:
Work on 2d electrostatic methods

RudolfWeeber committed:
Coulomb mixed periodicity test + reference data

Asserts are currently disabled and the test is not included in CMakeList, as the methods being tested don't agree with each other, yet.

RudolfWeeber committed:
coulomb 2d reference data generator (direct sumation w circular cutoff)

RudolfWeeber committed:
Merge branch 'python' of git://github.com/espressomd/espresso into electrostatic_mixed_periodicity

RudolfWeeber committed:
Merge commit 'refs/pull/1580/head' of git://github.com/espressomd/espresso into electrostatic_mixed_periodicity

RudolfWeeber committed:
Merge branch 'python' of git://github.com/espressomd/espresso into electrostatic_mixed_periodicity

RudolfWeeber committed:
Coulomb 2d test + core fixes

RudolfWeeber committed:
Prevent the use of mmm2d with Verlet lists

RudolfWeeber committed:
Merge branch 'python' of git://github.com/espressomd/espresso into electrostatic_mixed_periodicity

RudolfWeeber committed:
Include coulomb mixed periodicity test in cmake

KonradBreitsprecher committed:
elc vs mmm2d fixes(interface consistency, const_pot mpi bug, const_pot energy sign) + tests(inert, dielectric, const_pot, nonneutral); node_grid setter

KonradBreitsprecher committed:
merged python branch

KonradBreitsprecher committed:
Docu, samples and tutorials adapted to elc interface changes. Minor fix for elc_vs_mmm2d testcases

KonradBreitsprecher committed:
incode docu for elc, mmm2d sanity check for verlet lists. mmm2d does not work with verlet lists.

RudolfWeeber commented:
Could you please integrate this with the mixed periodicity
electrostatics test branch? Some changes seem to be in both.

KonradBreitsprecher committed:
sphinx errors

KonradBreitsprecher committed:
integrated rudolf/electrostatic_mixed_periodicity and fixed mixed periodicity test

KonradBreitsprecher committed:
merged python

KonradBreitsprecher committed:
Update elc.cpp

Added ELC_on_resort_particles() in ELC_init(), which might not be called otherwise in 'snapshot integration'.

fweik commented on src/python/espressomd/actors.pyx:
```diff
@@ -10,8 +10,7 @@ cdef class Actor(object):
                        MagnetostaticExtension=False,
                        HydrodynamicInteraction=False,
-                       ElectrostaticExtensions=False,
-                       Scafacos=True)
```
If the Scafacos here no longer needed?

KonradBreitsprecher replied:
That one came with @RudolfWeeber.

fweik replied:
Rudolf is looking into it.

fweik commented on src/core/mmm2d.cpp:
```diff
@@ -1010,7 +1013,7 @@ static void setup_PQ(int p, int q, double omega, double fac)
   double e, e_di_l, e_di_h;
   double *llclcblk;
-  double *lclimgebot=nullptr, *lclimgetop=nullptr;
+  double *lclimgebot=NULL, *lclimgetop=NULL;
```
nullptr

fweik commented on src/python/espressomd/cellsystem.pyx:
```diff
@@ -199,7 +199,17 @@ cdef class CellSystem(object):
+                node_grid[2] = _node_grid[2]
+                mpi_err = mpi_bcast_parameter(FIELD_NODEGRID)
+                handle_errors("mpi_bcast_parameter failed")
+                #raise Exception('node_grid is not settable by the user.')
```
commented out code

fweik commented on src/python/espressomd/electrostatic_extensions.pyx:
```diff
@@ -30,6 +30,57 @@ IF ELECTROSTATICS and P3M:
+                                  algorithm and will throw an error.
+        far_cut                 : float, optional
+                                  Cut off radius, use with care, intended for testing purposes. 
+        """
```
iirc there should be an empty line at the end of the docstring.

KonradBreitsprecher committed:
review flo

KonradBreitsprecher committed:
Merge branch 'elstat2d_tests' of https://github.com/KonradBreitsprecher/espresso into elstat2d_tests

RudolfWeeber committed:
Re-add scafacos as there-can-only-be-one actor

RudolfWeeber commented:
When I build this with scafacos, the coulomb cloud wall test segfaults at exit.

RudolfWeeber committed:
Only copy coulomb/dipolar energy from gpu if methods are active

RudolfWeeber committed:
Return to smaller gap and box for elc, fix p2nfft 2d periodic test

RudolfWeeber commented:
@KonradBreitsprecher @fweik : 
* fixed the segfault (gpu energies being copied to memory that wasn't there)
* reduced elc gap size (your fix fixed the overconfident tuning and it work swith a cubic box and gap size=1)
* Adjusted the field accuracy in the p2nfft test for it to produce sufficient accuracy
* reactivated Scafacos as a there-can-only-be-one actor
Please review (and merge) the branch pr1638 at my gihtub. Then, this pr can be merged, from my side.

KonradBreitsprecher commented:
Looks good. Tested again with scafacos and got no issues.

fweik commented:
So ready to merge?

RudolfWeeber commented:
On Mon, Dec 04, 2017 at 03:46:24PM +0000, Florian Weik wrote:
> So ready to merge?
Yes.

RudolfWeeber committed:
Merge branch 'python' of git://github.com/espressomd/espresso into pr1638

RudolfWeeber committed:
testsuite: Work around array write-protection

RudolfWeeber committed:
Scafacos: Work on r_cut tuning in parallel sims

RudolfWeeber committed:
Work on Scafacos tuning deadlocks

RudolfWeeber committed:
Merge branch 'python' of git://github.com/espressomd/espresso into pr1638

RudolfWeeber committed:
sort out scafacos parallel tuning issues

RudolfWeeber commented:
I pushed fixes for parallel scafacos electrostatics. 
@fweik, if you are happy with the code and the CI tests pass, I think, this is ready.

RudolfWeeber commented:
I tested with SCAFACOS and with SCAFACOS_DIPOLES coulomb_cloud_wall _mixed_periodicity dipolar_mdlc_scafacos scafacos_1d2d
